### ISSUE TITLE
[BUG] WithPfxCertificateToUpload doesn't appear to actually upload, WithExistingCertificate doesn't work outside of the initial ResourceGroup

### ISSUE THREAD

**Issue by carsontwilson** [Azure/azure-libraries-for-net#850]
Imagine you have two resource groups, Group A and Group B and within those groups are various app services.

If you upload a new private PFX to group A, and then try to apply that same PFX to group B using the thumbprint _or_  cert name you will get a null reference exception.

This is NOT a restriction if you use the portal. If you go to the app service in Group B in the Portal, and try to apply the certificate there, it will appear in the dropdown and you can select it. In fact, this will work for _any_ app service in any group within the subscription.

This is also not a restriction in PowerShell's AzureRm. Provided you uploaded the cert to one App Service (i.e. one in group A), you can then apply it using the thumbprint to _any_ app service as long as it's in the same subscription (the portal as the same restriction).

I expected the Azure SDK to behave the same as PowerShell if not the portal. This is convenient when you spin up and down a lot of app services (we have hundreds), because you don't have to include the cert or the password in the cert application process.

**Comment by carsontwilson** [comment]
Additionally, if you look at the PFX itself by opening it from the PFX tab on the TLS/SSL Settings page it shows the same cert, with the same ID, in the same resource as the one uploaded in Group A, even though I am forced to upload it again in Group B.

That suggests it's not really using the uploaded certificate, but still the certificate uploaded in the original group as one would except from the Portal / Powershell behaviour.

**Comment by carsontwilson** [comment]
I think I can just about confirm it's not really uploading anything. I did another test on a separate subscription.

If you try the following:

`WebAppRm.Update()
                .DefineSslBinding()
                    .ForHostname(hostName)
                    .WithPfxCertificateToUpload(ceritificatePath, certificatePassword)
                    .WithSniBasedSsl()
                    .Attach()
                .Apply();`

It will fail. You must upload the cert through the Portal on one resource on that sub _anywhere_ then you can use the "upload" functionality in the scenario above.

If you don't, you'll get an error back stating something along the lines of object not found (I'll get the precise error message).

So, to recap:

1. You must upload the cert to any resource through the portal, otherwise the cert update will fail _despite_ calling the upload cert command
2. Once you have done #1, you cannot reference the cert by ThumbPrint  app services in resources other than the resource where the cert was originally updated... you _must_ "upload" the cert again
3. It does not appear you are actually uploading the cert at all, despite what the method is suggesting. This is evidenced by the issue in #1, and the fact that when you inspect the certificate through the portal that is "uploaded" each time, the reference is always to the first cert uploaded, even though you are supposedly uploading it for every web app.

Let me know if I can clarify further!

**Comment by carsontwilson** [comment]
Looks like there's on more thing - if you do the upload as in #1 in the previous remark, and then you try to reference it in another region it will not work. This is true for both Portal and PowerShell.

You must manually upload it, through the Portal, in each region and then you'll be able to proceed as above.

So, to be clear, you cannot use WithPfxCertificateToUpload to upload certs... anywhere (only works after you've manually uploaded through the portal for the specific region).

**Comment by weidongxu-microsoft** [comment]
@carsontwilson 

As WithPfxCertificateToUpload, we have a sample here, which works OK (I checked right now).
https://github.com/Azure/azure-libraries-for-net/blob/master/Samples/AppService/ManageWebAppWithDomainSsl.cs

I will check WithExistingCertificate a bit later.

**Comment by carsontwilson** [comment]
Hmmm - well we have 5 subscriptions representing different production and non production environments and the behaviour was the same for each sub. I had to update certs for nearly 300 sites today so there were many opportunities to reproduce it. 

One difference between your example and mine is you’re using and azure purchased domain where I am not. I’m not sure if that’s relevant but it definitely was not a one off.

**Comment by weidongxu-microsoft** [comment]
@carsontwilson 

My guess is it might more about the PFX then the azure domain / 3rd party domain.

I am trying to reproduce. A few questions:

1. What is your SDK version?
2. And does it have anything special about the certificate?

**Comment by ChenTanyi** [comment]
@carsontwilson 

Is it possible for you to test it with an azure purchased domain? (I think this can just use less than $0.01) Or I could provide a domain and SSL certificate for you to test.

I have tested the function app in two different resource group with same SSL certificate (from LetsEncrypt) and it works. I cannot reproduce it. Does it appear randomly?

```
https://www.teststar.biz/api/function1?name=abc&&code=U3l9Ee5dIoGsthNJadOkmn5JBrjK0feJFlAM4ei4nR98YWoGABz6Fw==

https://aaa.teststar.biz/api/Function1?code=xP/AQ9CNGUUaRa9zI6o9sayMBOj9gtltulOfxOka3Ly5kSS5ip6r/Q==&&name=456
```

**Comment by carsontwilson** [comment]
I'm on 1.26 right now, so I can update to 1.27. I have one more sub with no certs on it, so I can try there with 1.27 and see if it works. 

For our cert, I'm not sure if it would be considered special or not, but it is a domain verified wildcard cert purchased through a major cert reseller. 

Another difference is I don't have this line of code in mine: .WithManagedHostnameBindings(domain) (or this which would be more relevant for my case): WithThirdPartyHostnameBinding

My code looks like this:
WebAppRm.Update()
                .DefineSslBinding()
                    .ForHostname(hostName)
                    .WithPfxCertificateToUpload(ceritificatePath, certificatePassword)
                    .WithSniBasedSsl()
                    .Attach()
                .Apply();

I'm not sure if that's relevant or not...

**Comment by ChenTanyi** [comment]
@carsontwilson My test succeeds to use WithThirdPartyHostnameBinding. Here is my test steps:

1. Purchase a domain on Azure: **teststar.biz** (I think azure purchased domain is the same as others. It is just used for DNS resolution)
2. Get SSL certificate from LetsEncrypt: ***.teststar.biz**
3. Create a function app using Visual Studio: **FunctionApp120191023095102** https://docs.microsoft.com/en-us/azure/azure-functions/functions-create-your-first-function-visual-studio
4. Make **aaa.teststar.biz** resolve to CNAME **FunctionApp120191023095102.website.net**
5. Run following SDK code with success:
```csharp
            var func = azure.AppServices.FunctionApps
                .GetByResourceGroup(rgName, "FunctionApp120191023095102");

            func.Update()
                .DefineHostnameBinding()
                    .WithThirdPartyDomain("teststar.biz")
                    .WithSubDomain("aaa")
                    .WithDnsRecordType(CustomHostNameDnsRecordType.CName)
                    .Attach()
                .DefineSslBinding()
                    .ForHostname("aaa.teststar.biz")
                    .WithPfxCertificateToUpload(pfxFile, "test")
                    .WithSniBasedSsl()
                    .Attach()
                .Apply();
```

**Comment by carsontwilson** [comment]
Ok thanks so much for the test steps. For what it’s worth I was using an app service not a function but I would be surprised if that mattered. 

You have a bit more code for the binding than I do, I wonder if that’s relevant? I will try reproducing with my (fewer) steps and then try again with your set of steps and resort back.

**Comment by carsontwilson** [comment]
I don't seem to be experiencing this behaviour anymore... though I've done several updates of the packages since then.

I'm having a new issue with null reference exceptions though, I'll post that separately

**Comment by nickzhums** [comment]
@carsontwilson Since you are no longer experiencing this issue, We'll consider this one resolved and we can discuss your other issue (#917) separately. Thanks!
